### ISSUE TITLE
Feature Request - Possibly a Design Change

### ISSUE THREAD

**Issue by lenkawamoto** [Lixxia/fitbit-authenticator#8]
After using your application for a while, I've found that (most times) it takes a LONG time for the watch app to connect to the companion app and load the tokens and then display the TOTP codes.  Often times, the connection to the companion app fails or the loading of the tokens fails.  I am finding myself reaching for my phone and using the Authenticator+ mobile app to get my TOTP codes.

On my Pebble, it was nice to be able to launch the Authenticator app RATHER THAN the mobile app because my Pebble was A) on my wrist, B) didn't require authentication to unlock my phone, C) didn't require authentication to launch the mobile app.

Often times, I need my TOTP codes when I am in a server closet while my smartphone is sitting on my office desk attached to a charger.  The fact that the Pebble app was completely stand alone made it a VERY useful application to have on my wrist.

I know that you've mentioned that it takes a long time to do the SHA calculations, but (at least in my situation) I'd rather wait a couple of seconds per TOTP code and have it done by a completely standalone app than wait a couple of seconds only while near my smartphone only to encounter a loading (or communication) error and have to reach for my smartphone anyways.

Without knowing the Fitbit OS API specifics, here's what I propose:

On the Watch App:
1) Move the TOTP calculation back into the watch app.
2) On launch, don't establish a connection to the companion app.
3) The initial screen should ONLY display a list of the names of the stored tokens.
4) The user, then, taps on the name of the token for which (s)he would like the TOTP code generated.
5) display a new screen that ONLY shows the name of the token and the freshly calculated TOTP code, plus a time-out indicator -- perhaps the same border progress bars as you have right now.
6) This new screen is the ONLY place where the TOTP code is calculated.
7) store a HASH value of the currently stored tokens on the watch - to be used for SYNC purposes.
8) on the initial page (where you display the list of known tokens), have a "Sync Data" button (at the bottom?) to communicate with the companion app and get the latest token data (Send the HASH value to the companion app so IT can determine if an updated token list needs to be sent back to the watch app).
9) This communication (the "Sync Data") should ONLY retry a couple of times and gracefully fail if it can't establish connection -- in case the smartphone is out of range or some other communication issues arise.  This way the watch can still be used with its locally stored information.

On the Mobile App:
1) Anytime tokens are re-ordered, added, removed, calculate a new HASH value of the token list.
2) When the token list is updated and IF IT'S POSSIBLE, push the updated token data to the watch app. From what I observe in your application's behaviour, I suspect the companion app is NOT able to initiate a data push to the watch app.
3) Otherwise, when the Watch App requests updated token list information, use the HASH value to compare with what was passed by the Watch App to see if mobile app needs to send updated token data back to the watch app.

**Comment by ghost** [comment]
I haven't had any failed connections since I installed the latest update and am almost always within bluetooth distance when I need to authenticate, so I don't have the issues Len has had. On the other hand, I do like Len's ideas (it would be great to not have to worry about my phone dying and then not be able to authenticate). They make sense to me, also not knowing anything about the API, but I'm grateful for this app either way.

**Comment by Lixxia** [comment]
If it was just slow generation of tokens I'd definitely keep everything on the watch. I too came from the Pebble and it really bugs me to have to depend on the phone for faster calculation. That coupled with the spotty communication can definitely sour the experience.

However the main issue is that as you add more tokens the time to calculate each of them increases. At about 3 or 4 tokens the entire application crashes (I believe Fitbit automatically kills apps that are blocking for more than ~10seconds).

If I understand correctly with what you're proposing only one token would ever be calculated at a time (based on user selection) on the watch, which would avoid the crashing issue. I'll investigate and see if I'm able to implement a sort of "independent mode" that keeps things on the watch at the expense of displaying multiple tokens at the same time and with the understanding that calculation will take several seconds. Would this be an acceptable solution?

Really appreciate the feedback (both of you), and I've definitely had users report communication issues, and experienced them myself, so I can understand the frustration.

**Comment by lenkawamoto** [comment]
I think that limiting the calculations to just a single TOTP code (at a time) should help with the watch app usability... even if there's a slight delay in displaying the TOTP code.

Having the watch app operate without requiring the companion app would make the app more useful, in my opinion.

Currently, it's not quite as useful as I'd have hoped... and at times it's worse than had I reached for my smartphone (to use a mobile app) instead.

Thanks for being responsive to comments, and open to suggestions BTW.

**Comment by Lixxia** [comment]
I'm pretty frustrated at the limitations of the Versa regarding this, especially coming from the Pebble. I think the one token thing is probably the best solution until Fitbit can make some improvements to either their communication or expose an API for these kinds of calculations.

We just have to see if I can implement it without tearing up the code completely!

**Comment by lenkawamoto** [comment]
I had time to sleep on it and have some additional ideas.

Perhaps you don't need to spilt into multiple screens.

When the watch app is launched, only generate the TOTP code for the first listed item.  The other items would simply display the name.

(apologies for my beautiful ASCII art)

------------------------
aa bb cc
Github
------------------------
Google
------------------------
User@host
------------------------
[ Sync Tokens ]
------------------------


The user could, then, tap on another list item to have the watch app generate the TOTP code for IT.  Again, the app only calculates the TOTP code for only a single token.

------------------------
Github
------------------------
Google
------------------------
dd ee ff
User@host
------------------------
[ Sync Tokens ]
------------------------

The watch app does all of this without communicating to the companion app.

The user taps on the "Sync Tokens" to initiate communications to the companion app.  Again, I think it's important for the process to fail gracefully and allow for the watch app to be used even if the Sync process fails.

Finally, I previously suggested storing a hash value for checking the Sync, but perhaps you can get by with storing a time stamp instead and can even display that below the Sync button (Last Sync: yyyy-mm-dd hh:mm:ss tz)

This might help you to integrate these changes without significantly changing your code, at least in respects to the UI.  I think it also simplifies the UX.

**Comment by Lixxia** [comment]
Hey! Thought I'd give an update on this. I implemented it the way you described in your most recent post, but there are still a lot of bugs to iron out. I'm still not sure if the response time is livable, it blocks the entire UI for ~3-6 seconds every time it's updated (every 30 seconds, upon user click, or upon app open). If the app is running in the background and the user receives a notification, waking the screen will attempt to generate a token (thus causing a delay on receiving the notification). 

Additionally if the user happens to wake the app, or tap a new token right before the 30 second threshold is hit, that's going to cause a total of ~12 seconds of frozen UI before the calculations are resolved (since it's trying to do 2).

I'd like to hear anyones input on the usability of this. One solution could be that the token is only updated upon tapping, perhaps with the timer running passively in the background to signify when an update should be initiated. However this kind of defeats the purpose in my eyes (requires manual user input for every new token).

Also apologies for the delays on this, work has been busy so I haven't had a lot of time!

**Comment by lenkawamoto** [comment]
Thanks so much for working on implementing this.  I really appreciate it, and it sounds like there are others interested in your app.

I could see how the current scheme might make the app unusable and I like your suggested solution (generate the auth code only when tapped).

Is it important that your app automatically updates the auth code when the timer expires?  I think not.

Certainly, it'd be handy if the app could automatically refresh with new codes, but we know the Versa is poorly equipped to handle the calculations.  I think tapping an entry to get a new auth code calculated is probably more tolerable than the potential ~12 second frozen UI.  Especially if it means I don't need to be tethered to a companion app on my phone.

Here's what I envision:
1. User launches the app
2. The app immediately is able to list all entries for which it can generate auth codes.  Each entry has an empty space for the 6 digit auth code to be displayed, and perhaps space on the far right of its entry line for an "expiration" icon.  Below the last entry item is a "Sync" button to sync the entry list with the companion app.
3. No auth codes are currently calculated nor displayed... no timer is running.  By NOT automatically calculating the auth code for the first entry, you allow the user to more quickly select an entry that is NOT the first item on the list.
4. The user taps on an entry to get an auth code generated.
5. The app now shows a spinner while it calculates the auth code, the UI is effectively frozen at this time for approximately 3-6 seconds (as you indicated above).
6. Once the calculation is done, the entry line is updated with the auth code (in "AA BB CC" or "AAA BBB" pattern as configured in the settings), the background timer is started, and the spinner is removed.
7. When the timer expires, the entry line is updated with an "expired" icon but the auth code is left displayed.  Some systems can allow for various deltas from the expiration times -- since the app is not automatically refreshing with new codes, it should keep the stale auth codes visible in case they're still acceptable.

When do you remove the stale auth code, though?  Do you just keep displaying them until the user taps to generate a new auth code or shuts down the app?

What to do if the user taps on another entry before the timer expires?  Can you have multiple timer threads?  Should you keep the auth code for the previous entry displayed?  With or Without an "expire" indicator?  Should you remove the auth code from the previous entry line?  I think the ideal scenario is that you can have multiple timer threads (or you handle multiple time out values with a single background timer) so that you can update each entry line with an "expire" indicator as needed.

Similarly, what to do if the user taps on another entry and all timers have expired?  Should you keep the previous entry displayed? etc...

Again, thank you for working on this application.

**Comment by Lixxia** [comment]
Hm, yes by removing the auto update every 30 seconds I could technically allow multiple "threads" since there's less of a computation risk as long as the user doesn't go crazy tapping every token.

I'm thinking display the token when tapped, then change the color to red after it's expired but keep it displayed for an additional, let's say 10 seconds? Then clear it. Gives the user sufficient warning/extra usage I think.

I'll play around with this and see what I get.

Thanks for the feedback!

**Comment by Lixxia** [comment]
After a long time away from working on this I've finally updated! It's not quite what we discussed but after the new Fitbit 3.0 SDK there's enough of a performance improvement to shift everything back to the watch without crashing. It will support all 10 tokens and takes roughly half a second to calculate each token, resulting in a (hopefully) tolerable delay of up to around 6 seconds. I've submitted the update today (3/30) and hopefully it will be available for download sometime next week.

If the delay becomes too irritating I can look back into single-token support.
